File: Venkatesh/Huffman-Coding/Huffman.java
import java.util.PriorityQueue;
import java.util.HashMap;
import java.util.Map;

public class Huffman {
    private static class HuffmanNode implements Comparable<HuffmanNode> {
        char ch;
        int freq;
        HuffmanNode left, right;

        HuffmanNode(char ch, int freq, HuffmanNode left, HuffmanNode right) {
            this.ch = ch;
            this.freq = freq;
            this.left = left;
            this.right = right;
        }

        @Override
        public int compareTo(HuffmanNode that) {
            return this.freq - that.freq;
        }
    }

    public static HuffmanNode buildHuffmanTree(Map<Character, Integer> freq) {
        PriorityQueue<HuffmanNode> pq = new PriorityQueue<>();
        for (Map.Entry<Character, Integer> entry : freq.entrySet()) {
            pq.add(new HuffmanNode(entry.getKey(), entry.getValue(), null, null));
        }

        while (pq.size() > 1) {
            HuffmanNode left = pq.poll();
            HuffmanNode right = pq.poll();
            pq.add(new HuffmanNode('\0', left.freq + right.freq, left, right));
        }

        return pq.poll();
    }

    public static void generateHuffmanCode(HuffmanNode node, String prefix, Map<Character, String> codebook) {
        if (node != null) {
            if (node.ch != '\0') {
                codebook.put(node.ch, prefix);
            } else {
                generateHuffmanCode(node.left, prefix + "0", codebook);
                generateHuffmanCode(node.right, prefix + "1", codebook);
            }
        }
    }

    public static String huffmanEncode(String str, Map<Character, String> codebook) {
        StringBuilder encoded = new StringBuilder();
        for (char ch : str.toCharArray()) {
            encoded.append(codebook.get(ch));
        }
        return encoded.toString();
    }

    public static String huffmanDecode(String str, HuffmanNode root) {
        StringBuilder decoded = new StringBuilder();
        HuffmanNode node = root;
        for (int i = 0; i < str.length(); i++) {
            node = str.charAt(i) == '0' ? node.left : node.right;
            if (node.ch != '\0') {
                decoded.append(node.ch);
                node = root;
            }
        }
        return decoded.toString();
    }

    public static void main(String[] args) {
        String str = "hello world";

        Map<Character, Integer> freq = new HashMap<>();
        for (char ch : str.toCharArray()) {
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }

        HuffmanNode root = buildHuffmanTree(freq);
        Map<Character, String> codebook = new HashMap<>();
        generateHuffmanCode(root, "", codebook);
        String encoded = huffmanEncode(str, codebook);
        String decoded = huffmanDecode(encoded, root);

        System.out.println("Encoded: " + encoded);
        System.out.println("Decoded: " + decoded);
    }
}

File: Venkatesh/Huffman-Coding/Huffman.py
import heapq
from collections import Counter, namedtuple

class HuffmanNode(namedtuple("HuffmanNode", ["char", "freq", "left", "right"])):
    def __lt__(self, other):
        return self.freq < other.freq

def build_huffman_tree(freq):
    heap = [HuffmanNode(char, freq, None, None) for char, freq in freq.items()]
    heapq.heapify(heap)
    
    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)
        node = HuffmanNode(None, left.freq + right.freq, left, right)
        heapq.heappush(heap, node)
    
    return heap[0]

def generate_huffman_codes(node, prefix="", codebook={}):
    if node.char is not None:
        codebook[node.char] = prefix
    else:
        generate_huffman_codes(node.left, prefix + "0", codebook)
        generate_huffman_codes(node.right, prefix + "1", codebook)
    return codebook

def huffman_encode(str, codebook):
    return ''.join(codebook[char] for char in str)

def huffman_decode(str, root):
    node, decoded = root, []
    for bit in str:
        node = node.left if bit == '0' else node.right
        if node.char is not None:
            decoded.append(node.char)
            node = root
    return ''.join(decoded)

str = "hello world"
freq = Counter(str)
root = build_huffman_tree(freq)
codebook = generate_huffman_codes(root)
encoded_data = huffman_encode(str, codebook)
decoded_data = huffman_decode(encoded_data, root)
print(f"Encoded: {encoded_data}")
print(f"Decoded: {decoded_data}")